install/install.go

```diff
@@ -156,93 +155,97 @@ func FromRepo(ctx context.Context, pi goolib.PackageInfo, repo, cache string, rm
+		return err
 	}
-	zs, err := extractSpec(arg)
+	zs, err := extractSpec(pkgPath)
 	if err != nil {
-		return nil, fmt.Errorf("error extracting spec file: %v", err)
+		return fmt.Errorf("error extracting spec file: %v", err)
 	}
-	if !ri {
-		ni, err := NeedsInstallation(goolib.PackageInfo{Name: zs.Name, Arch: zs.Arch, Ver: zs.Version}, *state)
+	if !shouldReinstall {
+		ni, err := NeedsInstallation(goolib.PackageInfo{Name: zs.Name, Arch: zs.Arch, Ver: zs.Version}, db)
 		if err != nil {
-			return nil, err
+			return err
 		}
 		if !ni {
 			fmt.Printf("%s.%s.%s or a newer version is already installed on the system\n", zs.Name, zs.Arch, zs.Version)
-			return nil, nil
+			return nil
 		}
 	}
 
-	logger.Infof("Starting install of %q, version %q from %q", zs.Name, zs.Version, arg)
+	logger.Infof("Starting install of %q, version %q from %q", zs.Name, zs.Version, pkgPath)
 	fmt.Printf("Installing %s %s...\n", zs.Name, zs.Version)
 
-	if err := resolveConflicts(zs, state); err != nil {
-		return nil, err
+	if err := resolveConflicts(zs, db); err != nil {
+		return err
 	}
 	for p, ver := range zs.PkgDependencies {
 		pi := goolib.PkgNameSplit(p)
-		mi, err := minInstalled(goolib.PackageInfo{Name: pi.Name, Arch: pi.Arch, Ver: ver}, *state)
-		if err != nil {
-			return nil, err
-		}
-		if mi {
+		if ok, err := minInstalled(goolib.PackageInfo{Name: pi.Name, Arch: pi.Arch, Ver: ver}, db); err != nil {
+			return err
+		} else if ok {
 			logger.Infof("Dependency met: %s.%s with version greater than %s installed", pi.Name, pi.Arch, ver)
 			continue
 		}
-		return nil, fmt.Errorf("package dependency %s %s (min version %s) not installed", pi.Name, pi.Arch, ver)
+		return fmt.Errorf("package dependency %s %s (min version %s) not installed", pi.Name, pi.Arch, ver)
 	}
 	for _, pkg := range zs.Replaces {
 		pi := goolib.PkgNameSplit(pkg)
-		ins, err := minInstalled(goolib.PackageInfo{Name: pi.Name, Arch: pi.Arch, Ver: pi.Ver}, *state)
+		ins, err := minInstalled(goolib.PackageInfo{Name: pi.Name, Arch: pi.Arch, Ver: pi.Ver}, db)
 		if err != nil {
-			return nil, err
+			return err
 		}
 		if ins {
-			return nil, fmt.Errorf("cannot install, replaces installed package, remove first then try installation again: %s", pi)
+			return fmt.Errorf("cannot install, replaces installed package, remove first then try installation again: %s", pi)
 		}
 	}
 
 	dst := filepath.Join(cache, goolib.PackageInfo{Name: zs.Name, Arch: zs.Arch, Ver: zs.Version}.PkgName())
-	if err := copyPkg(arg, dst); err != nil {
-		return nil, err
+	if err := copyPkg(pkgPath, dst); err != nil {
+		return err
 	}
 
 	insFiles, err := installPkg(dst, zs, dbOnly)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if ri {
+	if shouldReinstall {
 		logger.Infof("Reinstallation of %q, version %q completed", zs.Name, zs.Version)
 		fmt.Printf("Reinstallation of %s completed\n", zs.Name)
-		return nil, nil
+		return db.AddPkg(client.PackageState{
```

Comment:
Instead of duplicating this code, use the if-else to handle logging for the reinstall vs install case, then let them both fall through to the same db.AddPkg code at the end of the function.